transient/sshfs.py

```diff
@@ -115,127 +108,92 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
+    sshfs_command = f"sudo -E sshfs {sshfs_options} :{local_dir} {remote_dir}"
+
+    # Because sshfs monopolizes stdout, progress markers go to stderr
+    ssh_command = f"""
+        set -e
+        echo TRANSIENT_SSH_COMPLETE >&2
+        sudo mkdir -p {remote_dir}
+        {sshfs_command}
+    """
+
     sshfs_config = ssh_config.override(
-        args=["-A", "-T", "-o", "LogLevel=ERROR",] + ssh_config.args
+        args=["-T", "-o", "LogLevel=ERROR"] + ssh_config.args
+    )
+    client = ssh.SshClient(sshfs_config, command=ssh_command)
+
+    logging.info(f"Sending sshfs mount command '{sshfs_command}'")
+    ssh_proc = client.connect(
+        timeout=connect_timeout,
+        stdin=sftp_proc.stdout,
+        stdout=sftp_proc.stdin,
+        stderr=subprocess.PIPE,
+        tie_to_thread=False,
     )
-    client = ssh.SshClient(sshfs_config)
-    conn = client.connect_piped(timeout=connect_timeout)
-
-    try:
-        sshfs_options = "-o StrictHostKeyChecking=no -o allow_other"
-        sshfs_command = f"sudo -E sshfs {sshfs_options} {local_user}@10.0.2.2:{local_dir} {remote_dir}"
-
-        logging.info(f"Sending sshfs mount command '{sshfs_command}'")
-
-        sshfs_timeout = _SSHFS_MAX_RUN_TIME
-        if is_slow is True:
-            sshfs_timeout = _SSHFS_MAX_RUN_TIME_SLOW
-
-        # This is somewhat gnarly. The core of the issue is that sshfs is a FUSE mount,
-        # so it runs as a process (that gets backgrounded by default). SSH won't close
-        # the connection on its side until "it encounters end-of-file (eof) on the pipes
-        # connecting  to the stdout and stderr of the user program". This typically means
-        # you can do something like 'nohup <cmd> >/dev/null </dev/null 2>&1 &' to close
-        # all handles and ignore any hang ups. However, this doesn't work for SSHFS, as
-        # it spawns other processes that (I guess?) still have an open handle.
-        #
-        # This causes the SSH connetion to hang forever after the logout. Therefore, we
-        # need to close the connection on our end. So the trick here is to wait for some
-        # max time for this process to be done, then inspect the stdout for a sentinel
-        # value indicating that we _did_ get to the point where it should be OK to
-        # terminate the connection.
-        #
-        # See http://www.snailbook.com/faq/background-jobs.auto.html for some more info.
-        _, raw_stderr = conn.communicate(
-            input=f"""
-          set -e
-          sudo mkdir -p {remote_dir}
-          {sshfs_command}
-          echo TRANSIENT_SSHFS_DONE
-          exit
-        """.encode(
-                "utf-8"
-            ),
-            timeout=sshfs_timeout,
-        )
 
-        # Ensure returncode is set
-        conn.poll()
+    # Everything from here on out simply verifies that nothing went wrong.
+    if ssh_proc.stderr is not None:
+        first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
```

Comment:
Actually, I think it would be much better to have this be a timed blocking read (via `communicate`). This should also eliminate my concern below about something else writing to stderr. What would be ideal here would be a line buffered read with a timeout. That way we could set the timeout be be high, because it would almost always return almost immediately when the guest writes the magic line. However, I'm not sure it's possible to do that without more hacks than I'm willing to have.

Reply:
This should not hang. SSH will either timeout or get connection refused, and SSHFS is not yet called at this point.
This also acts as a dividing line between waiting for SSH, which takes a variable amount of time and will timeout on its own based on the user's arguments, and waiting for SSHFS, which is very fast and needs its own timeout. Using `communicate()` would always block for the maximum time unless something breaks, which would slow things down.